### ISSUE TITLE
@cc_on is slow and breaks jquery-migrate

### ISSUE THREAD

**Issue by meelash** [kayahr/console-shim#7]
Please see the discussion here: http://bugs.jquery.com/ticket/13274

Paul Irish recommends using this: https://gist.github.com/padolsey/527683 instead of @cc_on because it is slow.

**Comment by dmp42** [comment]
The reason for why "it" (or anything else using conditional compilation) breaks jquery is because they use @ in comments (which I don't quite get why it's not stripped out by minification, but that's a different topic). 
Anyhow, @cc_off should prevent any such side effect, and should definitely be done.

As for manipulating the DOM instead of using @cc:
- I have never seen numbers showing that cc is slow(er that manipulating the DOM) - can you share your source / tests cases?
- manipulating the DOM early is generally a dangerous idea, likely to cause problems
- the code section that uses cc is evaluated only once - even if that would be slow, the benefit is likely to be insignificant

**Comment by meelash** [comment]
It's not that @ is in the comments, it's a special keyword, added after minification for creating source maps for the browser debugger: http://net.tutsplus.com/tutorials/tools-and-tips/source-maps-101/

@cc_off does indeed solve the problem, but doesn't it make sense for you to do that at the end of your code, rather than forcing other people to add it to their code for defensive purposes? Also, @cc_off affects following code.

Regarding cc being slower, I don't have any numbers, just quoting from that link. I've asked for clarification there. But from my understanding, it's not a question of the actually cc call being slow, but rather the use of cc forces turning off js optimizations therefore causing all future evaluations to be slower. Perhaps paul will share some info about this with us.

**Comment by meelash** [comment]
Anyway, at least for me, the problem was solved by the jquery-migrate crew by removing source maps, which apparently were being generated by mistake, anyway.

**Comment by dmp42** [comment]
"@cc_off does indeed solve the problem, but doesn't it make sense for you to do that at the end of your code, rather than forcing other people to add it to their code for defensive purposes?"

Absolutely. This should be fixed in console-shim code.

"Also, @cc_off affects following code."

It's quite hard to find proper documentation about conditional compilation. Do you by any chance have further links/infos about cc_off behavior in IE, or cc impact on IE perf?
Even MSDN doesn't seem to have references about cc_off :( (http://msdn.microsoft.com/en-US/library/ahx1z4fs(v=vs.100).aspx)

Thanks a lot!

**Comment by kayahr** [comment]
Can you please give me an example how to use @cc_off? I can't find any documentation about this or any example and when I try to use it in IE myself I just get syntax errors or no effect (conditional compilation is still active)

**Comment by dmp42** [comment]
Have you tried:

if (window["**consoleShimTest**"] != null || 
    eval("/_@cc_on @_jscript_version <= 9 @cc_off @_/"))
{

That's supposed to be the syntax for console.

Put otherwise eval(/_@cc_on doSomething(); @cc_off @_/)

**Comment by kayahr** [comment]
Doesn't work here. I still can get the @_jscript_version after @cc_off. That's what I mean, it has no effect. Maybe it depends on IE version? Tested on IE9 standard mode. I guess it's safest to use the other solution instead (Creating DOM elements with conditional comments). This solution doesn't work in IE10 but in this case it doesn't matter because I have to check for IE <=9 anyway.

**Comment by kayahr** [comment]
By the way: It looks like IE is caching expression results. So if you run @_jscript_version BEFORE @cc_on and then trying again with the same expression AFTER @cc_on then you get no result. Example:

/_@_jscript_version@_/  => No result because CC is off
/_@cc_on@_/
/_@_jscript_version@_/ => Still no result (Most likely cached)
/_@_jscript_version @_/ => Now outputs the JScript version, the whitespace helped.

So to test if @cc_off works I did this:

/_@_jscript_version < 7 @_/   => No result because CC is off
/_@cc_on@_/
/_@_jscript_version < 8 @_/   => Return boolean result
/_@cc_off@_/
/_@_jscript_version < 9 @_/   => Still returns boolean result, so CC is still active

So it looks to me that this @cc_off is a myth. I can also run /_@cc_deactivateThisStinkingMess@_/ with the same result.

**Comment by dmp42** [comment]
"So it looks to me that this @cc_off is a myth."

Or existed in previous versions of IE and has been removed.

What a mess :/

**Comment by meelash** [comment]
Agreed about the mess. ;) It seems like it only exists in the memory of people who've used this feature back in the day and I can't find any reference to it in the docs either.

**Comment by evil-shrike** [comment]
This conditional compilation code (`eval("/*@cc_on @_jscript_version <= 9@*/"))`) doesn't work correctly in a case of higher version IE emulates lower version. For example, if in IE11 we switch document mode to IE8 then that `eval` return `false` and console methods won't be wrapped. 

It seems the cc check can be easily replaced with `document.documentMode < 9`.
